File: 347-top-k-frequent-elements/top-k-frequent-elements.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        int n = nums.size();
        vector<vector<int>>buckets(n+1);
        unordered_map<int,int>freq;
        vector<int>ans;
        for(int num : nums){
            freq[num]++;
        }

        for(auto f : freq){
            buckets[f.second].push_back(f.first);
        }

        for(int i = n; i>=0; i--){
            for(auto num : buckets[i]){
                ans.push_back(num);
                if(ans.size() == k) return ans;
            }
        }
        return {};
    }
};

File: 347-top-k-frequent-elements/top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer,Integer> freq = new HashMap<>();

        for(int num  : nums){
            freq.put(num, freq.getOrDefault(num,0) + 1);
        }
         
        PriorityQueue<int[]> pq = new PriorityQueue<>((a,b)->a[1] - b[1]);
        
        for(int key : freq.keySet()){
            pq.add(new int[]{key,freq.get(key)});
            if(pq.size() > k){
                pq.poll();
            }
        }

        int[] result = new int[k];
        int i = 0;

        while(!pq.isEmpty()){
            result[i++] = pq.poll()[0];
        } 

        return result;
    }
}